### PR TITLE
Remove jQuery dependency

### PR DESCRIPTION
**Pull request by richardvenneman**
Thanks for the great lib! We've been using Ahoy extensively in multiple projects.

In a recent project, only Ahoy.js depends on jQuery. Since I noticed that "Remove jQuery dependency" is listed as a TODO I thought I'd give it a shot.

This pull request is a starting point to start off the conversation. I would really like some input and assistance in removing the jQuery dependency.

Some things worth noting:

1. I replaced $.ajax with window.fetch. This will probably require the user to load a [Fetch polyfill](https://github.com/github/fetch/blob/master/fetch.js). Another solution would be to load another small AJAX library or write our own AJAX functions. Personally I think fetch is quite elegant and see that it's already being polyfilled quite a lot.
2. I replaced jQuery $() ready function with `DOMContentLoaded` event listeners. This works in IE9+,  not sure if stone age browser support is required here.
3. We only use Ahoy.js with `ahoy_matey` to create visits and do not use any of its event tracking functionality, so that will need some testing.

### CODE REVIEW

**Comment by ankane** [comment]
Hey Richard, this is pretty awesome. 👍 

The biggest hurdle for removing jQuery is the need for automated testing on different platforms and browsers. I've used [BrowserStack](https://www.browserstack.com/) in the past for manual testing, but unsure what people typically use for automated cross-browser JavaScript testing these days.

Aside from that, I think we can support AJAX requests without fetch pretty easily. Here's how we did it with [Chartkick](https://github.com/ankane/chartkick.js/commit/c7a2f315c7908e75446d97ed076fb8171b7a969d).

Unfortunately, I don't have much time to dedicate to this. I'm happy to give feedback, but you or someone else would need to get testing set up before I'd be comfortable switching to this.

Another option would be to experimentally support an `ahoy.standalone.js` library with a subset of browser support that people could use at their own risk. If we go this approach, I can help test events with the `ahoy_matey` gem.

**Comment by richardvenneman** [comment]
Thanks Andrew, I appreciate the feedback.

I think the Chartkick XHR approach will definitely do fine in this case.

Setting up a test suite would be great. I've come to love the simplicity of [Tape](https://github.com/substack/tape) and would be glad to setup a test suite. It might be better to do that in a separate pull request against the current version of ahoy.js so we can compare test output.

With regard to cross browser testing: I think a lot projects use [Sauce Labs](http://saucelabs.com) or [Testling](https://ci.testling.com) nowadays. This should be pretty easy to setup once we have a test suite.

I'll be abroad for the coming week but will contribute when I get back!

**Comment by ankane** [comment]
Sounds good. I agree with the approach of adding a test suite to the current version first to compare. I haven't used Tape or Sauce Labs before, but it looks like Tape is fairly popular and Sauce Labs has a free plan for open source projects, so sounds like a good combo.

**Comment by dts** [comment]
I'm in support of this!  I think depending on fetch is probably too soon (neither Safari nor IE support it yet).  XHR is totally reasonable for the few ways in which ahoy.js interacts with HTTP.  While fetch is awesome, depending on a polyfill defeats the purpose of dropping the dependency on jquery 😄 .  FWIW, the way I was about to shim this was to create a `ahoy-jq` file that implemented only the methods Ahoy needed, so you can either require jquery & ahoy, or ahoy-jq and Ahoy.

**Comment by richardvenneman** [comment]
I rebased master, so we now have a failing test suite 😄 

Thanks for the feedback! Yes, let's drop the Fetch polyfill and use a simple Chartkick-like XHR function. And by completely dropping the jQuery dependency we would only have to maintain a single version 😃

**Comment by davidguthu** [comment]
Excited to see this merge, is there anything that you guys need help with?

**Comment by richardvenneman** [comment]
I had some trouble with the changes after rebasing so I cherry-picked some commits on top of the master branch. I also [replaced the Fetch usage](https://github.com/ankane/ahoy.js/pull/10/commits/ff8ba399c4b7f2a467a694c48e8dac50a4e88aaa) with a [Chartkick](https://github.com/ankane/chartkick.js/commit/c7a2f315c7908e75446d97ed076fb8171b7a969d) inspired AJAX function.

Tests are passing locally but it seems Travis can't run them via Saucelabs:

> Zuul tried to run tests in saucelabs, however no saucelabs credentials were provided.
See the zuul wiki (https://github.com/defunctzombie/zuul/wiki/Cloud-testing) for info on how to setup cloud testing.

The `.travis.yml` seems to be configured correctly as described in the [Zuul wiki](https://github.com/defunctzombie/zuul/wiki/Travis-ci) 🤔 Maybe @ankane can help out here?

@davidguthu I think some help battle-testing this on a real website would be greatly appreciated!

**Comment by davidguthu** [comment]
I think the saucelabs error is related to encrypted credentials https://github.com/defunctzombie/zuul/issues/86

**Comment by davidguthu** [comment]
I deleted my prior post, thought I was being daft. It seems like this might be related to permissions https://github.com/defunctzombie/zuul/issues/86

**Comment by ankane** [comment]
Hey @richardvenneman, try syncing `.travis.yml` with master to fix the credentials issue.

**Comment by richardvenneman** [comment]
@ankane from the defunctzombie/zuul#86 issue:

> It is because the contributor is not allowed to run things with encrypted env vars. This is so that they can't just dump your credentials by just making a PR that does console.log in the tests once the env var is loaded. I don't really know a good way around this. It is a travis "feature" and not specific to zuul.

It is unfortunate, but we have two options:

1. Don't run Travis/SauceLabs on PR
2. Use unencrypted SauceLabs credentials (I think only if you're not using your SauceLabs account for other projects, or just create a new SauceLabs account specifically for this repo)

File: test/template.html [hunk 1]
```diff
@@ -1,6 +1,3 @@
-<meta content="authenticity_token" name="csrf-param" />
```
Comment (ankane) [inline test/template.html]: Why this change?

File: package.json [hunk 2]
```diff
@@ -39,6 +36,15 @@
     "test:local": "zuul --local 8080 test/ahoy_test.js"
   },
   "browserify": {
-    "transform": [ ["babelify", { "presets": [ "es2015" ] } ] ]
+    "transform": [
```
Comment (ankane) [inline package.json]: Let's keep the changeset to a minimum.

**Comment by ankane** [comment]
Hey @richardvenneman, looks like this is coming together :) I added a few comments inline.

File: ahoy.js [hunk 3]
```diff
@@ -345,27 +368,39 @@
...
-      var $target = $(e.currentTarget);
-      var properties = eventProperties(e);
-      properties.text = properties.tag == "input" ? $target.val() : $.trim($target.text().replace(/[\s\r\n]+/g, " "));
-      properties.href = $target.attr("href");
-      ahoy.track("$click", properties);
-    });
+    var elements = document.querySelectorAll("a, button, input[type=submit]");
+
+    for (var i = 0; i < elements.length; i++) {
+      elements[i].addEventListener('click', function (e) {
```
Comment (ankane) [inline ahoy.js]: Can we DRY up the `querySelectorAll`, loop over elements, `addEventListener` code?

**Comment by ankane** [comment]
Also, once the comments are resolved, I'll push a new branch so we can see the test results.

File: test/template.html (hunk 1 above)
Comment (richardvenneman) [inline test/template.html]: In ff8ba39 I replaced the dollar selector with a `document.head.querySelector`, therefor the csrf meta tags actually have to exist inside the head. I _thought_ this would not be a problem, but I now see that [`rails-ujs` uses a general `document.querySelctor`](https://github.com/rails/rails-ujs/blob/8a8a9c867e7a5c6ad0ab08abd5369ee9989464d7/src/utils/csrf.coffee) so that the csrf meta tags can appear outside of the head. Will revert this change.

File: package.json (hunk 2 above)
Comment (richardvenneman) [inline package.json]: I blame `npm uninstall jquery --save-dev`, but will revert this change!

File: ahoy.js (hunk 3 above)
Comment (richardvenneman) [inline ahoy.js]: Fixed in https://github.com/ankane/ahoy.js/pull/10/commits/c43d34dfecdb75d171d953a66922cf5be460499e but not entirely sure if this is what you meant. AFAIK this is the best way to [loop over a NodeList](https://developer.mozilla.org/nl/docs/Web/API/NodeList#Example) and it is not possible to [addEventListener on a NodeList](http://stackoverflow.com/questions/12362256/addeventlistener-on-nodelist). Just let me know what you think!

File: ahoy.js [hunk 4]
```diff
@@ -344,28 +369,44 @@
     ahoy.track("$view", properties);
   };
 
-  ahoy.trackClicks = function () {
-    $(document).on("click", "a, button, input[type=submit]", function (e) {
-      var $target = $(e.currentTarget);
+  ahoy.trackClick = function (element) {
```
Comment (ankane) [inline ahoy.js]: Let's hold off on external API changes in this PR.

File: ahoy.js (hunk 3 above)
Comment (ankane) [inline ahoy.js]: The code looks good, but the same pattern is repeated multiple times. I'd create an `onEvent(selector, eventName)` function to reduce duplicate code (just as jQuery has the `on` function). The number of lines in the `trackSubmits`, `trackChanges`, etc functions should stay constant.

Comment (richardvenneman) [inline ahoy.js]: Aah now I get it, thanks for the feedback!

**Comment by richardvenneman** [comment]
@ankane I just pushed some changes based on your feedback. Let me know what you think! 😄

**Comment by ankane** [comment]
This is coming together nicely. I pushed a `no-jquery` branch with tests passing. However, when adding IE 8, the tests hang. https://travis-ci.org/ankane/ahoy.js/jobs/216620984

I believe it's due to the issues mentioned here, but doesn't look like there's a fix: https://github.com/avocode/combokeys/issues/40

**Comment by richardvenneman** [comment]
Thanks for digging in @ankane. It seems the easiest for now is to switch to a different HTTP mocking lib, such as https://github.com/node-nock/nock. Would you like me to commit this change in this or a new pull request?

**Comment by ankane** [comment]
Sounds good, either works for me.

**Comment by richardvenneman** [comment]
I ended up replacing Sinon with [faux-jax](https://github.com/algolia/faux-jax) (which is maintained by algolia). It appears Nock only works well with Node and not the browser environment. Hope this one works with IE8!

**Comment by ankane** [comment]
Looks like `faux-jax` isn't working on IE 8 or 11 :(

https://travis-ci.org/ankane/ahoy.js/jobs/220343343

**Comment by richardvenneman** [comment]
🤦‍♂️ Maybe this is related to https://github.com/algolia/faux-jax/issues/26. Hopefully I'll have the time to dig in soon.

**Comment by jakeNiemiec** [comment]
Will this ever be merged? Is there anything to contribute?

**Comment by coryschires** [comment]
I would also love to see this get merged! 

Regarding IE8 support... Is it still necessary to support IE8? It's obviously very old and, AFAIK, it's not something most clients / companies are checking for these days. Maybe we could cut the IE8 requirement and merge as is?

If there's anything I can do to speed this along – including refinements to this PR – I'd be happy to pitch in.

**Comment by richardvenneman** [comment]
@jakeNiemiec @coryschires It's unfortunate but this pull seems to be hanging on IE8/IE11 support in tests, specifically AJAX mocks.

I was just reading the [Sinon.JS Fake XHR and server documentation](http://sinonjs.org/releases/v1.17.7/fake-xhr-and-server/) which states:

> When using the fake server in IE you also need sinon-ie.js. Load it after the first file.

I'll try to push these changes over the weekend!

**Comment by jakeNiemiec** [comment]
@richardvenneman That might be because IE <= 11 don't really support [`request.respond(200, { 'Content-Type': 'application/json' }, '{}');`](https://github.com/ankane/ahoy.js/blob/no-jquery/test/ahoy_test.js#L45)
image
image

**Comment by richardvenneman** [comment]
Thanks @jakeNiemiec, it definitely seems to be a problem with XHR but I'm not sure where to look exactly.

---

An update on this issue: I locally reverted the last commit, added the sinon-ie.js requirement and managed to get Saucelabs working. This is what I found:

IE8 doesn't pass because of the use of document DOMContentLoaded. So I guess we're going to need a wrapper function similar to `$(document).ready(function())`. This should be pretty easy. I also noted this in the pull request BTW:

> I replaced jQuery $() ready function with DOMContentLoaded event listeners. This works in IE9+, not sure if stone age browser support is required here.

After some digging it seems IE10 is failing because of an `open` call on a [superagent](https://github.com/visionmedia/superagent) client, which apparently is used by Zuul 😩

screen shot 2017-06-06 at 23 28 35

I'm not sure yet what to do about the IE10 issue.. I'm also a bit swamped currently 😨

**Comment by coryschires** [comment]
@richardvenneman Just a clarification. Is this a problem in the lib code **OR** is the problem limited to the test suite? 

If the problem is limited to the test suite, maybe we could consider merging and filing a separate issue for fixing the test suite in IE11?

**Comment by richardvenneman** [comment]
@coryschires both: IE8 fails because of the use of `DOMContentLoaded` (lib code), IE11 fails to something related tot the test suite.

@ankane is IE8 support still required? If so, I will add a custom `document.ready` function.

**Comment by ankane** [comment]
@richardvenneman yeah, still need IE8 support.

**Comment by jakeNiemiec** [comment]
_`January 11, 2016` - [How the credit card industry is helping the adoption of modern web standards ](https://blog.silverorange.com/how-the-credit-card-industry-is-helping-the-adoption-of-modern-web-standards-ec46c333605e)_

>"The PCI DSS mandates that all new applications support a minimum of TLSv1.1 by **June of 2016**. All applications, old or new must support TLSv1.1 by **June of 2018**.".

Customers on IE(`10 - older`), `Windows Xp` & `Vista` are falling out of PCI compliance. No **new** work/apps can support these platforms. [The deadline for old and new apps had already been delayed once ( original deadline was June 2016)](https://www.pcisecuritystandards.org/pdfs/15_12_18_SSL_Webinar_Press_Release_FINAL_(002).pdf).

The day after this article: _[Microsoft announced Support for older versions of Internet Explorer ended](https://www.microsoft.com/en-us/WindowsForBusiness/End-of-IE-support) (10 - older)_

A year later, Chrome surges, IE collapses: _January 7, 2017 - [2016 sees Internet Explorer usage collapse](https://arstechnica.com/information-technology/2017/01/2016-on-the-web-firefox-fights-back-as-microsofts-share-slumps/)_ 

TL;DR:
- Developers creating new e-commerce apps are currently (legally) compelled to block users on IE(`10 - older`), `Windows Xp` & `Vista`.
- Developers of existing apps have about a year left (June of 2018) to do the same.

@ankane, I recommend dropping IE(`10 - older`) support and releasing ahoy `0.3.0`.

**Comment by ankane** [comment]
@jakeNiemiec IE 8 still has over 1% market share. You're welcome to use this branch if you don't need to support it, but Ahoy won't be dropping support right now.

**Comment by richardvenneman** [comment]
I just pushed a wrapper function that should reintroduce IE8 support.

Spun up some VM's and tried getting everything to pass in all IEs but to no avail, unfortunately 😞

**Comment by victorarias** [comment]
Hello! @richardvenneman this may be unrelated to your problems, but I could only make your test suite pass consistently on some browsers (like IE11) after this change: https://github.com/Futurelearn/ahoy.js/commit/e6a639ccba251e473b6211aa8d80e5131aa62cad

By the way, since we don't support IE8 at FutureLearn, we have been using your changes since last week on a fork from Ahoy :tada: Thanks for the great work!

**Comment by richardvenneman** [comment]
@victorarias Thanks! Great find 😄  I knew something was up with that test but couldn't find the cause. Will implement this soon.

**Comment by richardvenneman** [comment]
OK with the latest fix incorporated (credit to @victorarias) the tests pass consistently on IE9 & 10. There is still one remaining issue with IE8. I tracked it down to this [`Object.defineProperty`](https://github.com/nodejs/readable-stream/blob/d6c391d7101946c8b8b97914fc46fd3322c450d1/lib/_stream_readable.js#L190) call in readable-stream. AFAIK this package is a node JS internal module. So I tried to downgrade Node (with nvm) to a version before [this code was added](https://github.com/nodejs/readable-stream/commit/98296bb6265b87aa3bbf2903532377844f5509ef#diff-ba6a0df0f5212f5cba5ca5179e209a17R188); node version 7.10.0. But unfortunately this didn't yield any results (because I think a vendored version is a dependency of something?)

So the current state is that is that IE8 crashes on a call in the readable-stream module, but I honestly don't know where to go from here with my current nodeJS knowledge.. 🆘 

screen shot 2017-07-14 at 22 09 19

**Comment by EmCousin** [comment]
Hi @richardvenneman, we are in the same case that @victorarias and since it is still not merged we used your branch. Thanks for the awesome work, it works like a charm!

Unfortunately my knowledge of NodeJS is too limited to help you on the last failing test on IE8, but if I ever get insights I won't hesitate to share them here ;)

**Comment by ankane** [comment]
The time has come... if all goes well, Ahoy 2.0 will ship today or tomorrow with this branch 🎉  If someone is able to confirm or add IE 8 support, that'd be great. Otherwise, I'll look at it after everything is ready to go on the Ruby library. Big thanks @richardvenneman for all your work on this!
